File: packages/gbnf-experiment/data/20260909T173101Z_8e3da990/ported_implementation/utils/errors/input_parse_error.py
from __future__ import annotations

from typing import TYPE_CHECKING

from .build_error_position import build_error_position
from .get_input_as_string import get_input_as_string

if TYPE_CHECKING:
    from ...grammar_graph.types import ValidInput

INPUT_PARSER_ERROR_HEADER_MESSAGE = 'Failed to parse input string:'


class InputParseError(Exception):
    def __init__(self, most_recent_input: "ValidInput", pos: int, previous_input: "ValidInput" = ''):
        self._most_recent_input = most_recent_input
        self._pos = pos
        self._previous_input = previous_input
        super().__init__('\n'.join([
            INPUT_PARSER_ERROR_HEADER_MESSAGE,
            '',
            *build_error_position(
                f'{get_input_as_string(previous_input)}{get_input_as_string(most_recent_input)}',
                pos + len(get_input_as_string(previous_input)),
            ),
        ]))
        self.name = 'InputParseError'

    @property
    def message(self) -> str:
        return str(self)

    @property
    def src(self) -> str:
        return f'{get_input_as_string(self._previous_input)}{get_input_as_string(self._most_recent_input)}'

    @property
    def error_for_most_recent_input(self) -> str:
        return '\n'.join([
            INPUT_PARSER_ERROR_HEADER_MESSAGE,
            '',
            *build_error_position(get_input_as_string(self._most_recent_input), self._pos),
        ])

    # JS-name alias for parity with the reference implementation.
    @property
    def errorForMostRecentInput(self) -> str:
        return self.error_for_most_recent_input

File: packages/gbnf-experiment/data/20260909T175143Z_62b61540/ported_implementation/src/utils/errors/input-parse-error.ts
import { buildErrorPosition } from './build-error-position';
import { ValidInput } from './errors-types';
import { getInputAsString } from './get-input-as-string';

export const INPUT_PARSER_ERROR_HEADER_MESSAGE = 'Failed to parse input string:';

const buildMessage = (src: string, pos: number): string => [
  INPUT_PARSER_ERROR_HEADER_MESSAGE,
  '',
  ...buildErrorPosition(src, pos),
].join('\n');

export class InputParseError extends Error {
  mostRecentInput: ValidInput;
  pos: number;
  previousInput: ValidInput;

  constructor(mostRecentInput: ValidInput, pos: number, previousInput: ValidInput = '') {
    super(buildMessage(
      `${getInputAsString(previousInput)}${getInputAsString(mostRecentInput)}`,
      pos + getInputAsString(previousInput).length,
    ));
    Object.setPrototypeOf(this, InputParseError.prototype);
    this.name = 'InputParseError';
    this.mostRecentInput = mostRecentInput;
    this.pos = pos;
    this.previousInput = previousInput;
  }

  get src(): string {
    return `${getInputAsString(this.previousInput)}${getInputAsString(this.mostRecentInput)}`;
  }

  get errorForMostRecentInput(): string {
    return buildMessage(getInputAsString(this.mostRecentInput), this.pos);
  }
}
